src/server/tiered_storage.cc

```diff
@@ -355,6 +365,73 @@ void TieredStorage::Free(PrimeIterator it, DbTableStats* stats) {
   stats->tiered_size -= len;
 }
 
+size_t TieredStorage::GetBinSize(size_t blob_len) {
+  CHECK(blob_len <= kMaxSmallBin);
+  unsigned bin_index = SmallToBin(blob_len);
+  return kSmallBins[bin_index];
+}
+
+void TieredStorage::Defrag(DbIndex db_index) {
+  // going through all the candidates that have been inserted from the Free() function
+  // the Free() function is the only source which reduced page bin utilization.
+  for (; !pages_to_defrag_.empty(); pages_to_defrag_.pop()) {
+    unsigned int offs_page = pages_to_defrag_.front();
+    // check if this page still exists
+    auto page_it = page_refcnt_.find(offs_page);
+    if (page_it == page_refcnt_.end())
+      continue;
+
+    // the page exists, now check if this is still under utilized as this page
```

Comment:
if pages_to_defrag_ was a set you would remove from it on free and will not need to do this checks